kythe/go/serving/xrefs/columnar/columnar_encoding.go

```diff
@@ -0,0 +1,366 @@
+}
+
+// Columnar file decorations group numbers.
+// See: kythe/proto/xref_serving.proto
+const (
+	columnarDecorationsIndexGroup              = -1 // no group number
+	columnarDecorationsTextGroup               = 0
+	columnarDecorationsTargetGroup             = 10
+	columnarDecorationsTargetOverrideGroup     = 20
+	columnarDecorationsTargetNodeGroup         = 30
+	columnarDecorationsTargetDefinitionGroup   = 40
+	columnarDecorationsDefinitionLocationGroup = 50
+	columnarDecorationsOverrideGroup           = 60
+	columnarDecorationsDiagnosticGroup         = 70
+)
+
+// KV is a single columnar key-value entry.
+type KV struct{ Key, Value []byte }
+
+// EncodeDecorationsEntry encodes a columnar FileDecorations entry.
+func EncodeDecorationsEntry(keyPrefix []byte, fd *xspb.FileDecorations) (*KV, error) {
+	switch e := fd.Entry.(type) {
+	case *xspb.FileDecorations_Index_:
+		return encodeDecorIndex(keyPrefix, fd.File, e.Index)
+	case *xspb.FileDecorations_Text_:
+		return encodeDecorText(keyPrefix, fd.File, e.Text)
+	case *xspb.FileDecorations_Target_:
+		return encodeDecorTarget(keyPrefix, fd.File, e.Target)
+	case *xspb.FileDecorations_TargetOverride_:
+		return encodeDecorTargetOverride(keyPrefix, fd.File, e.TargetOverride)
+	case *xspb.FileDecorations_TargetNode_:
+		return encodeDecorTargetNode(keyPrefix, fd.File, e.TargetNode)
+	case *xspb.FileDecorations_TargetDefinition_:
+		return encodeDecorTargetDefinition(keyPrefix, fd.File, e.TargetDefinition)
+	case *xspb.FileDecorations_DefinitionLocation_:
+		return encodeDecorDefinitionLocation(keyPrefix, fd.File, e.DefinitionLocation)
+	case *xspb.FileDecorations_Override_:
+		return encodeDecorOverride(keyPrefix, fd.File, e.Override)
+	case *xspb.FileDecorations_Diagnostic_:
+		return encodeDecorDiagnostic(keyPrefix, fd.File, e.Diagnostic)
+	default:
+		return nil, fmt.Errorf("unknown FileDecorations entry: %T", e)
+	}
+}
+
+func encodeDecorIndex(prefix []byte, file *spb.VName, idx *xspb.FileDecorations_Index) (*KV, error) {
+	key, err := keys.Append(prefix, file)
+	if err != nil {
+		return nil, err
+	}
+	val, err := proto.Marshal(idx)
+	if err != nil {
+		return nil, err
+	}
+	return &KV{key, val}, nil
+}
+
+func encodeDecorText(prefix []byte, file *spb.VName, t *xspb.FileDecorations_Text) (*KV, error) {
+	key, err := keys.Append(prefix, file, columnarDecorationsTextGroup, t.StartOffset, t.EndOffset)
+	if err != nil {
+		return nil, err
+	}
+	val, err := proto.Marshal(&xspb.FileDecorations_Text{Text: t.Text})
+	if err != nil {
+		return nil, err
+	}
+	return &KV{key, val}, nil
+}
+
+func encodeDecorTarget(prefix []byte, file *spb.VName, t *xspb.FileDecorations_Target) (*KV, error) {
+	key, err := keys.Append(prefix, file,
+		columnarDecorationsTargetGroup, t.StartOffset, t.EndOffset,
+		t.GetGenericKind(), int32(t.GetKytheKind()), t.Target)
+	if err != nil {
+		return nil, err
+	}
+
+	val, err := proto.Marshal(&xspb.FileDecorations_Target{})
```

Comment:
Why don't we need to marshal any content?

Reply:
It's all encoded in the key.